### ISSUE TITLE
Protocol incompatibility not detected when base class declared

### ISSUE THREAD

**Issue by indigoviolet** [microsoft/pyright#3636]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
When a protocol implementation declares the protocol as its base class, some errors disappear

**To Reproduce**
See code below.  If you change `class Foobar(SupportsBlah)` to `class Foobar:`, the correct error (below) is shown

```
Expression of type "Foobar" cannot be assigned to declared type "SupportsBlah"
  "Foobar" is incompatible with protocol "SupportsBlah"
    "foo" is an incompatible type
      Type "() -> dict[Unknown, Unknown]" cannot be assigned to type "() -> int"
        Function return type "dict[Unknown, Unknown]" is incompatible with type "int"
          "dict[Unknown, Unknown]" is incompatible with "int"
```

**Expected behavior**
Adding the base class should show the same error, probably in this case at the class declaration itself.

**Screenshots or Code**
If applicable, add screenshots or the text of the code (surrounded by triple back ticks) to help explain your problem.
```python
from typing import Protocol

class SupportsBlah(Protocol):
    def foo(self) -> int:
        ...

class Foobar(SupportsBlah):
    def foo(self) -> dict:
        return dict()


f: SupportsBlah = Foobar()

```

If your code relies on symbols that are imported from a third-party library, include the associated import statements and specify which versions of those libraries you have installed.

**VS Code extension or command-line**
vscode: v1.1.256

**Additional context**

**Comment by erictraut** [comment]
This is as designed. When a class explicitly derives from a protocol class, it is asserting that it implements the protocol. 

Pyright does have an optional check called `reportIncompatibleMethodOverride` that would catch this issue.

**Comment by indigoviolet** [comment]
Ok thanks. For context, I derived this expectation from " Explicitly including a protocol as a base class is also a way of documenting that your class implements a particular protocol, and it forces mypy to verify that your class implementation is actually compatible with the protocol" (https://mypy.readthedocs.io/en/stable/protocols.html#defining-subprotocols-and-subclassing-protocols). Of course, mypy and pyright are not expected to behave identically.

**Comment by erictraut** [comment]
Pyright does verify that if a protocol class doesn't provide an implementation for a method classes that derive from it must provide an implementation.

```python
from typing import Protocol

class SupportsBlah(Protocol):
    def foo(self) -> int:
        ...

class Foobar(SupportsBlah): # Error: Foobar does not implement "foo"
    ...
```

In your case, the base class does provide an implementation for this method. Method override compatibility is already validated as part of the `reportIncompatibleMethodOverride` check, so it's not included in protocol validation. If it were, you'd receive two error messages (one of them redundant) in this case, assuming that you enabled `reportIncompatibleMethodOverride`.

**Comment by erictraut** [comment]
Parameter names are not enforced for any dundered methods because this tends to produce a lot of false positives. Many of the core classes defined in the typeshed stdlib stubs define dundered methods with parameter names that are not marked as positional-only when they should be. There has been an effort to clean these up over time, and it has gotten a lot better, but I worry that if we remove this exemption, it will produce a lot of undesirable false positives for pyright users.

I suppose I could exempt `__call__` from this exemption, but I'm reluctant to apply a special case on top of a special case. That doesn't feel very principled. If we can come up with a defensible principle that doesn't create an undue burden for existing pyright users, I would be happy to apply it. Let me know if you have any suggestions.

**Comment by indigoviolet** [comment]
Thanks for the explanation. I don't currently have a principled suggestion - was just documenting the inconsistency and trying to understand the motivation. TBH, I don't care for the parameter name check elsewhere either -- I think that the parameter name should reflect the implementation's context/language, not the protocol's. I would turn it off if there were a configuration option for it.

Related:

```python
f: Foo = lambda a: 2
```
does throw the name check error (which I guess only means that the exemption for dunder methods is based on the method being checked, not what it is checked against).
